python/message.c

```diff
@@ -0,0 +1,1546 @@
+    return NULL;
+  }
+
+  PyUpb_CMessage_AssureWritable(self);
+  const upb_msgdef* msgdef = _PyUpb_CMessage_GetMsgdef(self);
+  const upb_filedef* file = upb_msgdef_file(msgdef);
+  const upb_extreg* extreg = upb_symtab_extreg(upb_filedef_symtab(file));
+  const upb_msglayout* layout = upb_msgdef_layout(msgdef);
+  upb_arena* arena = PyUpb_Arena_Get(self->arena);
+  upb_DecodeStatus status =
+      _upb_decode(buf, size, self->msg, layout, extreg, 0, arena);
+  Py_XDECREF(bytes);
+  if (status != kUpb_DecodeStatus_Ok) {
+    PyUpb_ModuleState* state = PyUpb_ModuleState_Get();
+    PyErr_Format(state->decode_error_class, "Error parsing message");
+    return NULL;
+  }
+  PyUpb_CMessage_SyncSubobjs(self);
+  return PyLong_FromSsize_t(size);
+}
+
+static PyObject* PyUpb_CMessage_Clear(PyUpb_CMessage* self, PyObject* args);
+
+static PyObject* PyUpb_CMessage_ParseFromString(PyObject* self, PyObject* arg) {
+  PyObject* tmp = PyUpb_CMessage_Clear((PyUpb_CMessage*)self, NULL);
+  Py_DECREF(tmp);
+  return PyUpb_CMessage_MergeFromString(self, arg);
+}
+
+static PyObject* PyUpb_CMessage_ByteSize(PyObject* self, PyObject* args) {
+  // At the moment upb does not have a "byte size" function, so we just
+  // serialize to string and get the size of the string.
+  PyObject* subargs = PyTuple_New(0);
+  PyObject* serialized = PyUpb_CMessage_SerializeToString(self, subargs, NULL);
+  Py_DECREF(subargs);
+  if (!serialized) return NULL;
+  size_t size = PyBytes_Size(serialized);
+  Py_DECREF(serialized);
+  return PyLong_FromSize_t(size);
+}
+
+static PyObject* PyUpb_CMessage_Clear(PyUpb_CMessage* self, PyObject* args) {
+  PyUpb_CMessage_AssureWritable(self);
+  const upb_msgdef* msgdef = _PyUpb_CMessage_GetMsgdef(self);
+  upb_msg_clear(self->msg, msgdef);
+  Py_RETURN_NONE;
+}
+
+static void PyUpb_CMessage_AbandonField(PyUpb_CMessage* self,
+                                        const upb_fielddef* f) {
+  if (self->unset_subobj_map && upb_fielddef_issubmsg(f)) {
+    PyObject* sub = PyUpb_WeakMap_Get(self->unset_subobj_map, f);
+    if (sub) {
+      PyUpb_CMessage_AssureWritable((PyUpb_CMessage*)sub);
+    }
+  }
+}
+
+static PyObject* PyUpb_CMessage_ClearExtension(PyUpb_CMessage* self,
+                                               PyObject* arg) {
+  PyUpb_CMessage_AssureWritable(self);
+  const upb_msgdef* msgdef = _PyUpb_CMessage_GetMsgdef(self);
+  const upb_fielddef* f = PyUpb_FieldDescriptor_GetDef(arg);
+  if (!f) return NULL;
+  if (upb_fielddef_containingtype(f) != msgdef) {
+    PyErr_Format(PyExc_ValueError, "Extension doesn't match (%s vs %s)",
+                 upb_msgdef_fullname(msgdef), upb_fielddef_fullname(f));
+  }
+  PyUpb_CMessage_AbandonField(self, f);
+  upb_msg_clearfield(self->msg, f);
+  Py_RETURN_NONE;
+}
+
+static PyObject* PyUpb_CMessage_ClearField(PyUpb_CMessage* self,
+                                           PyObject* arg) {
+  PyUpb_CMessage_AssureWritable(self);
```

Comment:
if it is already unset, why should we ensure it is writable?  Can't we simply bail out?

Reply:
can you add a comment to that effect?

Reply:
Done.